builder/dispatchers.go

```diff
@@ -226,10 +226,16 @@ func run(b *Builder, args []string, attributes map[string]bool, original string)
 	// set Cmd manually, this is special case only for Dockerfiles
 	b.Config.Cmd = config.Cmd
 	runconfig.Merge(b.Config, config)
+	// set build-time environment for 'run'. We let dockerfile
+	// environment override build-time environment.
+	env := b.Config.Env
```

Comment:
@duglin, you got it right. This proposal does introduce build time env-vars to setup the environment required for the RUN primitive which involves running a command, script etc while addressing portability concerns as raised in issue #4962 around effects of persisting such values in final/intermediate images. This proposal however doesn't change or address anything around notion of expansion of other variables/symbols that are/shall be supported by Dockerfile otherwise.

I think what you ask is support for the latter. While it technically sounds doable I would not like to overlap the notion of environment variables be mixed with variables/symbol for expansion since the two serve different purposes and might be better implemented as separate command-line options.

If there is a value add to have variables/symbols for expansion in other primitives and available through command-line, I would be happy to investigate. Let me know.

Hope this clarifies.

Just a side note, while looking at the code for handling ADD primitive (in file `builder/dispatchers.go`) I couldn't find the logic for variable expansion. Most likely I missed some code path, could you kindly point me to relevant code if you are aware, just to enhance my understanding.